linera-execution/src/wasm/common.rs

```diff
@@ -352,8 +274,59 @@ where
     ///     argument: &[u8],
     /// ) -> Result<Vec<u8>, ExecutionError>
     /// ```
-    pub fn handle_query(self, context: &QueryContext, argument: &[u8]) -> PollSender<Vec<u8>> {
-        GuestFutureActor::<A::HandleQuery, A>::spawn((*context, argument.to_owned()), self)
+    pub fn handle_query(
+        self,
+        context: &QueryContext,
+        argument: &[u8],
+    ) -> WasmResultFuture<Vec<u8>> {
+        let context = *context;
+        let argument = argument.to_owned();
+
+        self.run_wasm_guest(move |application, store| {
+            application.handle_query(store, context, argument)
+        })
+    }
+}
+
+impl<A> WasmRuntimeContext<A>
+where
+    A: ApplicationRuntimeContext + Send + Unpin + 'static,
+{
+    /// Spawns a thread to execute a Wasm guest.
+    fn run_wasm_guest<T>(
+        mut self,
+        guest_operation: impl FnOnce(&A, &mut A::Store) -> Result<Result<T, String>, A::Error>
+            + Send
+            + 'static,
+    ) -> WasmResultFuture<T>
+    where
+        T: Send + 'static,
+    {
+        let (result_sender, result_receiver) = oneshot::channel();
+
+        // TODO(#925): Replace the host future queue with a read-write-lock.
+        if let Some(future_queue) = self.future_queue.take() {
+            tokio::spawn(future_queue.collect::<()>());
+        }
+
+        thread::spawn(move || {
```

Comment:
Do we have a task for merging the threads of a transaction?

Reply:
Yes, I added a link in the code. Thanks!